src/Sparrow/Json/JsonContextPoolBase.cs

```diff
@@ -10,129 +9,53 @@
-        private static ObjectPool<ContextStack> _contextStackPool = new ObjectPool<ContextStack>(() => new ContextStack());
+        private static readonly ObjectPool<ContextStack> _contextStackPool = new ObjectPool<ContextStack>(() => new ContextStack());
+
+        [ThreadStatic]
+        private static ContextStackThreadReleaser _releaser;
 
         private class ContextStackThreadReleaser
         {
-            private HashSet<JsonContextPoolBase<T>> _parents = new HashSet<JsonContextPoolBase<T>>();
-            int _threadId;
-            public List<ThreadIdHolder> ThreadIdHolders = new List<ThreadIdHolder>();            
+            private readonly int _ownerThread;
+            private readonly StrongReference<ConcurrentDictionary<int, ContextStack>> _contextStacksByThreadId;
 
-            public ContextStackThreadReleaser()
+            public ContextStackThreadReleaser(int ownerThread, StrongReference<ConcurrentDictionary<int, ContextStack>> contextStacksByThreadId)
             {
-                _threadId = NativeMemory.CurrentThreadStats.InternalId;
+                _ownerThread = ownerThread;
+                _contextStacksByThreadId = contextStacksByThreadId;
             }
 
             ~ContextStackThreadReleaser()
             {
-                foreach (var item in ThreadIdHolders)
-                {
-                    Debug.Assert(item.ThreadId == _threadId);
-                    item.ThreadId = -1;
-                }
-                // we remove the references from the thread dictionary
-                // it is possible that the pool is no longer referenced and was collected, and as such
-                // the finalizers for the context stack values were already run, if not, they will run soon
-                // anyway, so we just clear it.
-                foreach (var parent in _parents)
-                {
-                    parent._contextStacksByThreadId?.TryRemove(_threadId, out var contextStack);
-                }
-            }
-
-            public bool AddContextPool(JsonContextPoolBase<T> parent)
-            {
-                return _parents.Add(parent);
-            }
-
-            public void RemoveContextPool(JsonContextPoolBase<T> parent)
-            {
-                _parents.Remove(parent);
-            }
-
-            public void RemoveThreadIdHolder(ThreadIdHolder holder)
-            {
-                ThreadIdHolders.Remove(holder);
+                var current = _contextStacksByThreadId.Value;
+                if (current != null) // may have already been finalized
+                    current.TryRemove(_ownerThread, out _);
             }
         }
 
-        [ThreadStatic]
-        private static ContextStackThreadReleaser _releaser;
-        
-
         private void EnsureCurrentThreadContextWillBeReleased(int currentThreadId)
-        {            
+        {
             if (_releaser == null)
```

Comment:
If null, create it.
If not null, add yourself to it. But may already exists, so probably use a set.

Reply:
I preferred to add the strong ref instead of "myself" (this == pool).  Because the ref set to null at pool's Dispose, and then I do not care about this parent anyhow